[BUG] GeoTileGridIT.testGeoShapes failure

Seems to be reproducible:

```
./gradlew ':modules:geo:internalClusterTest' --tests "org.opensearch.geo.search.aggregations.bucket.GeoTileGridIT.testGeoShapes" -Dtests.seed=EC3F504EF22998E0

org.opensearch.geo.search.aggregations.bucket.GeoTileGridIT > testGeoShapes FAILED
    java.lang.AssertionError: Geotile 3/7/4 has wrong doc count  expected:<36> but was:<1>
        at __randomizedtesting.SeedInfo.seed([EC3F504EF22998E0:682E8817D55CBFD]:0)
        at org.junit.Assert.fail(Assert.java:89)
        at org.junit.Assert.failNotEquals(Assert.java:835)
        at org.junit.Assert.assertEquals(Assert.java:647)
        at org.opensearch.geo.search.aggregations.bucket.GeoTileGridIT.testGeoShapes(GeoTileGridIT.java:78)
```

https://build.ci.opensearch.org/job/gradle-check/10625/consoleFull

/cc @navneet1v

--- a/modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/bucket/AbstractGeoBucketAggregationIntegTest.java
+++ b/modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/bucket/AbstractGeoBucketAggregationIntegTest.java
@@ -40,7 +40,7 @@ import static org.opensearch.test.hamcrest.OpenSearchAssertions.assertAcked;
  */
 public abstract class AbstractGeoBucketAggregationIntegTest extends GeoModulePluginIntegTestCase {
 
-    protected static final int MAX_PRECISION_FOR_GEO_SHAPES_AGG_TESTING = 4;
+    protected static final int MAX_PRECISION_FOR_GEO_SHAPES_AGG_TESTING = 2;
 
     protected static final int MIN_PRECISION_WITHOUT_BB_AGGS = 2;
 
@@ -98,7 +98,7 @@ public abstract class AbstractGeoBucketAggregationIntegTest extends GeoModulePlu
             }
 
             i++;
-            final Set<String> values = generateBucketsForGeometry(geometry, geometryDocValue, isShapeIntersectingBB);
+            final Set<String> values = generateBucketsForGeometry(geometry, geometryDocValue);
             geoshapes.add(indexGeoShape(GEO_SHAPE_INDEX_NAME, geometry));
             for (final String hash : values) {
                 expectedDocsCountForGeoShapes.put(hash, expectedDocsCountForGeoShapes.getOrDefault(hash, 0) + 1);
@@ -112,16 +112,11 @@ public abstract class AbstractGeoBucketAggregationIntegTest extends GeoModulePlu
      * Returns a set of buckets for the shape at different precision level. Override this method for different bucket
      * aggregations.
      *
-     * @param geometry           {@link Geometry}
-     * @param geoShapeDocValue   {@link GeoShapeDocValue}
-     * @param intersectingWithBB boolean
+     * @param geometry         {@link Geometry}
+     * @param geoShapeDocValue {@link GeoShapeDocValue}
      * @return A {@link Set} of {@link String} which represents the buckets.
      */
-    protected abstract Set<String> generateBucketsForGeometry(
-        final Geometry geometry,
-        final GeoShapeDocValue geoShapeDocValue,
-        final boolean intersectingWithBB
-    );
+    protected abstract Set<String> generateBucketsForGeometry(final Geometry geometry, final GeoShapeDocValue geoShapeDocValue);
 
     /**
      * Prepares a GeoPoint index for testing the GeoPoint bucket aggregations. Different bucket aggregations can use

--- a/modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/bucket/GeoHashGridIT.java
+++ b/modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/bucket/GeoHashGridIT.java
@@ -268,18 +268,15 @@ public class GeoHashGridIT extends AbstractGeoBucketAggregationIntegTest {
     }
 
     @Override
-    protected Set<String> generateBucketsForGeometry(
-        final Geometry geometry,
-        final GeoShapeDocValue geometryDocValue,
-        boolean intersectingWithBB
-    ) {
+    protected Set<String> generateBucketsForGeometry(final Geometry geometry, final GeoShapeDocValue geometryDocValue) {
         final GeoPoint topLeft = new GeoPoint();
         final GeoPoint bottomRight = new GeoPoint();
         assert geometry != null;
         GeoBoundsHelper.updateBoundsForGeometry(geometry, topLeft, bottomRight);
         final Set<String> geoHashes = new HashSet<>();
+        final boolean isIntersectingWithBoundingRectangle = geometryDocValue.isIntersectingRectangle(boundingRectangleForGeoShapesAgg);
         for (int precision = MAX_PRECISION_FOR_GEO_SHAPES_AGG_TESTING; precision > 0; precision--) {
-            if (precision > MIN_PRECISION_WITHOUT_BB_AGGS && intersectingWithBB == false) {
+            if (precision > MIN_PRECISION_WITHOUT_BB_AGGS && isIntersectingWithBoundingRectangle == false) {
                 continue;
             }
             final GeoPoint topRight = new GeoPoint(topLeft.getLat(), bottomRight.getLon());

--- a/modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/bucket/GeoTileGridIT.java
+++ b/modules/geo/src/internalClusterTest/java/org/opensearch/geo/search/aggregations/bucket/GeoTileGridIT.java
@@ -134,20 +134,20 @@ public class GeoTileGridIT extends AbstractGeoBucketAggregationIntegTest {
      * Returns a set of buckets for the shape at different precision level. Override this method for different bucket
      * aggregations.
      *
-     * @param geometry           {@link Geometry}
-     * @param geoShapeDocValue   {@link GeoShapeDocValue}
-     * @param intersectingWithBB
+     * @param geometry         {@link Geometry}
+     * @param geoShapeDocValue {@link GeoShapeDocValue}
      * @return A {@link Set} of {@link String} which represents the buckets.
      */
     @Override
-    protected Set<String> generateBucketsForGeometry(Geometry geometry, GeoShapeDocValue geoShapeDocValue, boolean intersectingWithBB) {
+    protected Set<String> generateBucketsForGeometry(final Geometry geometry, final GeoShapeDocValue geoShapeDocValue) {
         final GeoPoint topLeft = new GeoPoint();
         final GeoPoint bottomRight = new GeoPoint();
         assert geometry != null;
         GeoBoundsHelper.updateBoundsForGeometry(geometry, topLeft, bottomRight);
         final Set<String> geoTiles = new HashSet<>();
+        final boolean isIntersectingWithBoundingRectangle = geoShapeDocValue.isIntersectingRectangle(boundingRectangleForGeoShapesAgg);
         for (int precision = MAX_PRECISION_FOR_GEO_SHAPES_AGG_TESTING; precision > 0; precision--) {
-            if (precision > MIN_PRECISION_WITHOUT_BB_AGGS && intersectingWithBB == false) {
+            if (precision > MIN_PRECISION_WITHOUT_BB_AGGS && isIntersectingWithBoundingRectangle == false) {
                 continue;
             }
             geoTiles.addAll(

--- a/modules/geo/src/main/java/org/opensearch/geo/search/aggregations/bucket/geogrid/cells/GeoShapeCellValues.java
+++ b/modules/geo/src/main/java/org/opensearch/geo/search/aggregations/bucket/geogrid/cells/GeoShapeCellValues.java
@@ -10,6 +10,7 @@ package org.opensearch.geo.search.aggregations.bucket.geogrid.cells;
 
 import org.opensearch.common.geo.GeoBoundingBox;
 import org.opensearch.common.geo.GeoShapeDocValue;
+import org.opensearch.geometry.Rectangle;
 import org.opensearch.index.fielddata.AbstractSortingNumericDocValues;
 import org.opensearch.index.fielddata.GeoShapeValue;
 
@@ -57,8 +58,7 @@ abstract class GeoShapeCellValues extends AbstractSortingNumericDocValues {
      * @opensearch.internal
      */
     static class BoundedCellValues extends GeoShapeCellValues {
-
-        private final GeoBoundingBox geoBoundingBox;
+        private final Rectangle geoBoundingBox;
 
         public BoundedCellValues(
             final GeoShapeValue geoShapeValue,
@@ -67,7 +67,7 @@ abstract class GeoShapeCellValues extends AbstractSortingNumericDocValues {
             final GeoBoundingBox boundingBox
         ) {
             super(geoShapeValue, precision, encoder);
-            this.geoBoundingBox = boundingBox;
+            this.geoBoundingBox = new Rectangle(boundingBox.left(), boundingBox.right(), boundingBox.top(), boundingBox.bottom());
         }
 
         /**
@@ -78,7 +78,7 @@ abstract class GeoShapeCellValues extends AbstractSortingNumericDocValues {
          */
         @Override
         void relateShape(final GeoShapeDocValue geoShapeDocValue) {
-            if (intersect(geoShapeDocValue.getBoundingRectangle())) {
+            if (geoShapeDocValue.isIntersectingRectangle(geoBoundingBox)) {
                 // now we know the shape is in the bounding rectangle, we need add them in longValues
                 // generate all grid that this shape intersects
                 final List<Long> encodedValues = encoder.encode(geoShapeDocValue, precision);
@@ -86,20 +86,11 @@ abstract class GeoShapeCellValues extends AbstractSortingNumericDocValues {
                 for (int i = 0; i < encodedValues.size(); i++) {
                     values[i] = encodedValues.get(i);
                 }
+            } else {
+                // As the shape is not intersecting with GeoBounding box, we need to reset the GeoShapeCellValues
+                // calling this function resets the CellValues for the current shape.
+                resize(0);
             }
-        }
-
-        /**
-         * Validate that shape is intersecting the bounding box provided as input.
-         *
-         * @param rectangle {@link GeoShapeDocValue.BoundingRectangle}
-         * @return true or false
-         */
-        private boolean intersect(final GeoShapeDocValue.BoundingRectangle rectangle) {
-            return geoBoundingBox.pointInBounds(rectangle.getMaxLongitude(), rectangle.getMaxLatitude())
-                || geoBoundingBox.pointInBounds(rectangle.getMaxLongitude(), rectangle.getMinLatitude())
-                || geoBoundingBox.pointInBounds(rectangle.getMinLongitude(), rectangle.getMaxLatitude())
-                || geoBoundingBox.pointInBounds(rectangle.getMinLongitude(), rectangle.getMinLatitude());
         }
 
     }

